File: src/main/java/dev/enderman/minecraft/plugins/projectiles/better/listener/FireballSpawnListener.java
package dev.enderman.minecraft.plugins.projectiles.better.listener;

import org.bukkit.NamespacedKey;
import org.bukkit.entity.Fireball;
import org.bukkit.entity.Ghast;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntitySpawnEvent;
import org.bukkit.persistence.PersistentDataContainer;
import org.bukkit.persistence.PersistentDataType;
import org.bukkit.projectiles.ProjectileSource;
import org.jetbrains.annotations.NotNull;

import dev.enderman.minecraft.plugins.projectiles.better.BetterProjectilesPlugin;

public class FireballSpawnListener implements Listener {

    private final BetterProjectilesPlugin plugin;

    public FireballSpawnListener(BetterProjectilesPlugin plugin) {
        this.plugin = plugin;
    }

    @EventHandler
    public void onFireBallSpawn(@NotNull EntitySpawnEvent event) {
        if (event.getEntity() instanceof Fireball fireball) {
            ProjectileSource source = fireball.getShooter();

            if (source instanceof Ghast ghast) {
                PersistentDataContainer container = ghast.getPersistentDataContainer();

                NamespacedKey nuclearGhastMobKey = plugin.getNuclearGhastMobKey();

                boolean isNuclearGhast = Boolean.TRUE.equals(container.get(nuclearGhastMobKey, PersistentDataType.BOOLEAN));

                if (isNuclearGhast) {

                }
            }
        }
    }
}

File: src/main/java/dev/enderman/minecraft/plugins/projectiles/better/listener/FireballSpawnListener.kt
package dev.enderman.minecraft.plugins.projectiles.better.listener

import dev.enderman.minecraft.plugins.projectiles.better.BetterProjectilesPlugin
import org.bukkit.entity.Fireball
import org.bukkit.entity.Ghast
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.ProjectileLaunchEvent
import org.bukkit.persistence.PersistentDataType

class FireballSpawnListener(private val plugin: BetterProjectilesPlugin) : Listener {

  @EventHandler
  fun onFireBallSpawn(event: ProjectileLaunchEvent) {
    val entity = event.entity

    if (entity is Fireball) {
      val source = entity.shooter

      if (source is Ghast) {
        val container = source.persistentDataContainer
        val nuclearGhastMobKey = plugin.nuclearGhastMobKey
        val isNuclearGhast = true == container.get(nuclearGhastMobKey, PersistentDataType.BOOLEAN)

        if (isNuclearGhast) {
          val projectileContainer = entity.persistentDataContainer
          projectileContainer.set(plugin.nuclearFireballKey, PersistentDataType.BOOLEAN, true)
        }
      }
    }
  }
}
